subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java

```diff
@@ -0,0 +1,301 @@
+import org.gradle.internal.fingerprint.impl.IgnoredPathFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.NameOnlyFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.OutputFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.RelativePathFileCollectionFingerprinter;
+import org.gradle.internal.hash.DefaultFileHasher;
+import org.gradle.internal.hash.FileHasher;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.StreamHasher;
+import org.gradle.internal.nativeintegration.filesystem.FileSystem;
+import org.gradle.internal.serialize.HashCodeSerializer;
+import org.gradle.internal.service.ServiceRegistration;
+import org.gradle.internal.snapshot.WellKnownFileLocations;
+import org.gradle.internal.vfs.VirtualFileSystem;
+import org.gradle.internal.vfs.impl.DefaultVirtualFileSystem;
+import org.gradle.internal.vfs.impl.RoutingVirtualFileSystem;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+
+public class VirtualFileSystemServices extends AbstractPluginServiceRegistry {
+    /**
+     * System property to enable feature preview for partial invalidation.
+     */
+    public static final String VFS_PARTIAL_INVALIDATION_ENABLED_PROPERTY = "org.gradle.unsafe.vfs.partial-invalidation";
+
+    /**
+     * System property to enable feature preview for retaining VFS state between builds.
+     */
+    public static final String VFS_RETENTION_ENABLED_PROPERTY = "org.gradle.unsafe.vfs.retention";
+
+    /**
+     * When retention is enabled, this system property can be used to pass a comma-separated
+     * list of file paths that have changed since the last build.
+     *
+     * @see #VFS_RETENTION_ENABLED_PROPERTY
+     */
+    public static final String VFS_CHANGES_SINCE_LAST_BUILD_PROPERTY = "org.gradle.unsafe.vfs.changes";
+
+    public static boolean isPartialInvalidationEnabled(Map<String, String> systemPropertiesArgs) {
+        return getSystemProperty(VFS_PARTIAL_INVALIDATION_ENABLED_PROPERTY, systemPropertiesArgs) != null;
+    }
+
+    public static boolean isRetentionEnabled(Map<String, String> systemPropertiesArgs) {
+        return getSystemProperty(VFS_RETENTION_ENABLED_PROPERTY, systemPropertiesArgs) != null;
+    }
+
+    public static Iterable<String> getChangedPathsSinceLastBuild(Map<String, String> systemPropertiesArgs) {
+        String changeList = getSystemProperty(VFS_CHANGES_SINCE_LAST_BUILD_PROPERTY, systemPropertiesArgs);
+        if (changeList == null) {
+            return ImmutableSet.of();
+        }
+        return Splitter.on(',')
+            .omitEmptyStrings()
+            .split(changeList);
+    }
+
+    @Nullable
+    private static String getSystemProperty(String systemProperty, Map<String, String> systemPropertiesArgs) {
+        return System.getProperty(systemProperty, systemPropertiesArgs.get(systemProperty));
+    }
```

Comment:
I am very unhappy about this solution. Let's take a look if we can inject something like a `SystemProperties` service that has the merged properties available.

Reply:
I wonder if the `systemPropertyArgs` should go first, so they would override whatever is set via system properties already.

Reply:
I changed it.